src/crab/wto.hpp

```diff
@@ -1,467 +1,180 @@
- * RECIPIENT'S SOLE REMEDY FOR ANY SUCH MATTER SHALL BE THE IMMEDIATE,
- * UNILATERAL TERMINATION OF THIS AGREEMENT.
- *
- ******************************************************************************/
-
+// Copyright (c) Prevail Verifier contributors.
+// SPDX-License-Identifier: MIT
 #pragma once
 
-#include <memory>
-#include <set>
+// This file implements Weak Topological Ordering as defined in
+// Bournacle, "Efficient chaotic iteration strategies with widenings", 1993
+// http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.38.3574
+//
+// Using the example from section 3.1 in the paper, the graph:
+//
+//         4 --> 5 <-> 6
+//         ^ \________ |
+//         |          vv
+//         3 <-------- 7
+//         ^           |
+//         |           v
+//   1 --> 2 --------> 8
+//
+// results in the WTO: 1 2 (3 4 (5 6) 7) 8
+// where a single vertex is represented via wto_vertex_t, and a
+// cycle such as (5 6) is represented via a wto_cycle_t.
+// Each arrow points to a wto_component_t, which can be either a
+// single vertex such as 8, or a cycle such as (5 6).
+
+#include <stack>
 #include <vector>
-#include <forward_list>
-#include <map>
-#include <variant>
-
-#include "crab/cfg.hpp"
-#include "crab/cfg_bgl.hpp"
-#include "crab/interval.hpp"
-#include "crab_utils/debug.hpp"
-
-namespace crab {
-
-
-using vertex_descriptor_t = typename boost::graph_traits<cfg_t>::vertex_descriptor;
-
-
-using out_edge_iterator_t = typename boost::graph_traits<cfg_t>::out_edge_iterator;
-
-
-using edge_descriptor_t = typename boost::graph_traits<cfg_t>::edge_descriptor;
-
-
-class wto_t;
-class wto_vertex_t;
-class wto_cycle_t;
-
-
-class wto_nesting_t final {
 
-    friend class wto_t;
-    friend class wto_vertex_t;
-    friend class wto_cycle_t;
+using label_t = crab::label_t;
+using wto_component_t = std::variant<std::shared_ptr<class wto_cycle_t>, label_t>;
+using wto_partition_t = std::vector<std::shared_ptr<wto_component_t>>;
 
-  private:
-    using node_list_t = std::vector<vertex_descriptor_t>;
-    using node_list_ptr = std::shared_ptr<node_list_t>;
-
-    node_list_ptr _nodes;
-
-  public:
-    using iterator = typename node_list_t::iterator;
-    using const_iterator = typename node_list_t::const_iterator;
-
-  private:
-    explicit wto_nesting_t(const node_list_ptr& l) : _nodes(std::make_shared<node_list_t>(*l)) {}
+#include "crab/cfg.hpp"
+#include "crab/wto_cycle.hpp"
+#include "crab/wto_nesting.hpp"
 
-    int compare(wto_nesting_t& other) const {
-        auto this_it = this->begin();
-        auto other_it = other.begin();
-        while (this_it != this->end()) {
-            if (other_it == other.end()) {
-                return 1;
-            } else if (*this_it == *other_it) {
-                ++this_it;
-                ++other_it;
+class wto_t final {
+    // Original control-flow graph.
+    const crab::cfg_t& _cfg;
+
+    std::map<label_t, int> _dfn;
+    int _num;
+    std::stack<label_t> _stack;
```

Comment:
Documentation for these fields would be nice.

Reply:
I agree, but the research paper does not define them (it doesn't even expand dfn as an acronym anywhere in the paper), so the only documentation I have is the name of the variable.

Reply:
Looks like Bourdoncle's [thesis](https://ucaf45ebca6288e2e58dc2397944.dl.dropboxusercontent.com/cd/0/inline2/BL1qBmxQ3GtQB9bm7YKk-2hR60OBbgjDxGcZlx-rSuoBvya7b_ibrCg79WzGcuGL0x-N0HI3T9NO1wf0VeUhM4wZK8uSEdbs0146OToEdu_KhT_640GiN_coE7JhMy-BauNWQv_2Oq81yOmzOKS09mxCRKkZV2trWG0cm1OeGc2MlSKx2qSbsv0D02L_Vq6HgQGbe_-ypkjVa-PhgIWj-mvsNuhQ9nDueUcFclJvqug1TJagz7niqsheEEbV8DqXoYOpV5QUabbJ_F6ax-8Rqeo_MOS0bsz6miguHlnvMPbcWahQoES2m-ONWPHrrSHHfmmVLxcmn1EaG5C-WJSQmz5bbpxkmaaegU0yWr-NLjqAPGL2vY0Jjo27YjEOeaFSMkg/file#) (almost all in French) expands DFN as "depth first number"